pkg/providers/ibmcloud.go

```diff
@@ -0,0 +1,331 @@
+		return "", errors.New("Can't find preset with name " + presetname)
+	}
+
+	orderkeys := []datatypes.Container_Product_Order_SshKeys{
+		{
+			SshKeyIds: []int{*key},
+		},
+	}
+
+	// Build a container_Product_Order object.
+	orderTemplate := datatypes.Container_Product_Order{
+		Quantity:         sl.Int(1),
+		Location:         sl.String(location),
+		PackageId:        sl.Int(packageId),
+		PresetId:         sl.Int(presetId),
+		Prices:           prices,
+		Hardware:         server,
+		UseHourlyPricing: sl.Bool(true),
+		SshKeys:          orderkeys,
+	}
+
+	// Get SoftLayer_Product_Order service.
+	service := services.GetProductOrderService(p.client)
+
+	//result, err := service.VerifyOrder(&orderTemplate)
+	_, err = service.PlaceOrder(&orderTemplate, sl.Bool(false))
+	if err != nil {
+		//if strings.Contains(err.Error(), "SoftLayer_Exception_Order_HardwareUnavailable") {
+		return "", err
+	}
+	return hostname, nil
+}
+
+func (p *ibmcloudProvider) Acquire(poolSize int, poolName string, poolType string) (Resource, error) {
+
+	res := Resource{}
+	id, err := p.serverCreate(poolName, p.config.Preset)
+	if err != nil {
+		return res, err
+	}
+
+	// Node don't appear in ibmcloud immediatly, Should we wait for the node to appear?
+	// would prevent a bunch of node not found errors in the logs from acquirecomplete
+
+	res.Id = id
+	return res, nil
+}
+
+func (p *ibmcloudProvider) getNodeByName(name string) (*datatypes.Hardware, error) {
+	service := services.GetAccountService(p.client)
+	nodes, err := service.Mask("id;hostname;primaryIpAddress;hardwareStatus;billingItem;hourlyBillingFlag").GetHardware()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, node := range nodes {
+		if *node.Hostname == name {
+			return &node, nil
+		}
+	}
+	return nil, errors.New("Node node found")
+}
+
+func (p *ibmcloudProvider) ensureSSHKey(newkey string) (*int, error) {
+	sshKeys, err := services.GetAccountService(p.client).GetSshKeys()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, key := range sshKeys {
+		if *key.Key == newkey {
+			return key.Id, nil
+		}
+	}
+	newKey := datatypes.Security_Ssh_Key{}
+	newKey.Key = sl.String(newkey)
+	// TODO: could do with something desciptive here (poolname if available)
+	newKey.Label = sl.String(strings.Split(uuid.New().String(), "-")[0])
+	newKey, err = services.GetSecuritySshKeyService(p.client).CreateObject(&newKey)
+
+	if err != nil {
+		return nil, err
+	}
+	return newKey.Id, nil
+}
+
+func (p *ibmcloudProvider) AcquireCompleted(id string) (bool, Resource, error) {
+
+	res := Resource{
+		Id: id,
+	}
+
+	node, err := p.getNodeByName(id)
+	if err != nil {
+		return false, res, err
+	}
+
+	// This may happen for a short period (20 seconds or so) after a node is initially created
+	if node == nil {
+		return false, res, fmt.Errorf("error getting node: %w", err)
+	}
+
+	// not ready yet
+	if *node.HardwareStatus.Status != "ACTIVE" {
+		return false, res, nil
+	}
+
+	res.Address = *node.PrimaryIpAddress
```

Comment:
could be moved after the `IsPortOpen` check?

Reply:
can do